v2/tools/generator/internal/reporting/type_catalog_report.go

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package reporting
+
+import (
+	"fmt"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/pkg/errors"
+	"io"
+	"os"
+	"sort"
+)
+
+type TypeCatalogReport struct {
+	title                  string
+	defs                   astmodel.TypeDefinitionSet
+	inlinedTypes           astmodel.TypeNameSet // Set of types that we inline when generating the report
+	optionIncludeFunctions bool
+}
+
+func NewTypeCatalogReport(defs astmodel.TypeDefinitionSet) *TypeCatalogReport {
+	return &TypeCatalogReport{
+		defs:         defs,
+		inlinedTypes: astmodel.NewTypeNameSet(),
+	}
+}
+
+// SaveTo writes the report to the specified file
+func (tcr *TypeCatalogReport) SaveTo(filePath string) error {
+
+	file, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		file.Close()
+
+		// if we are panicking, the file will be in a broken
+		// state, so remove it
+		if r := recover(); r != nil {
+			os.Remove(filePath)
+			panic(r)
+		}
+	}()
+
+	err = tcr.WriteTo(file)
+	if err != nil {
+		// cleanup in case of errors
+		file.Close()
```

Comment:
Do we need this block here? We'll handle this in defer function already?

Reply:
I think we do - the defer block will only remove the file if we're in mid-panic, whereas we always want to remove it if there was a problem.

Reply:
Correct me if I'm wrong - If we close here, and again close in defer block(before exitting), it'll panic anyway(as we're closing a closed file) and never return the error?

Reply:
The second call to `Close()` returns an error (which is ignored anyway); [it doesn't panic](https://pkg.go.dev/os#File.Close).

FWIW, I copied the structure from `go_source_file.go` (in our `astmodel` package), so I hope it's not a broken technique!